### PR TITLE
Add vector search param source

### PR DESCRIPTION
**Pull request by VijayanB**
### Description
Added new param source to partition vector dataset and
neighbors. This will be passed to runner to perform
search and compare response with neighbors for recall
calculation.

This param source extends Search ParamSource to inherit search's
other query parameters.
Vector Param Source will add additional paramter that are required
for vector serach operation type.

### Issues Resolved
Part of #103 

### Testing
- [x] New functionality includes testing

[Describe how this change was tested]
```
make test

tests/workload/params_test.py::VectorSearchParamSourceTests::test_invalid_data_set_format PASSED [ 96%]
tests/workload/params_test.py::VectorSearchParamSourceTests::test_invalid_data_set_path PASSED [ 96%]
tests/workload/params_test.py::VectorSearchParamSourceTests::test_missing_params PASSED [ 96%]
tests/workload/params_test.py::VectorSearchParamSourceTests::test_partition_bigann PASSED [ 96%]
tests/workload/params_test.py::VectorSearchParamSourceTests::test_partition_hdf5 PASSED [ 96%]
tests/workload/params_test.py::VectorSearchPartitionPartitionParamSourceTestCase::test_params PASSED [ 96%]


================= 1206 passed, 5 skipped, 3 warnings in 15.59s =================
```

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by VijayanB** [comment]
This depends on #424

**Comment by VijayanB** [comment]
Tested this PR with 1m-128d-l2-hdf5 vector dataset

File: osbenchmark/workload/params.py [hunk 1]
```diff
@@ -797,6 +799,183 @@ def params(self):
...
+        elif is_dataset_exhausted:
+            raise StopIteration
+        vector = self.data_set.read(1)[0]
+        neighbor = self.neighbors_data_set.read(1)[0]
+        true_neighbors = list(map(str, neighbor[:self.k]))
+        self.query_params.update({
+            "neighbors": true_neighbors,
+            "request-params": {
+                "_source": "false",
+                # we need to set it to true as this data source is used for actual queries
```
Comment (jmazanec15) [inline osbenchmark/workload/params.py:947]: This seems to contradict L948?

File: osbenchmark/workload/params.py [hunk 2]
```diff
@@ -797,6 +799,183 @@ def params(self):
...
+            self.current_rep += 1
+        elif is_dataset_exhausted:
+            raise StopIteration
+        vector = self.data_set.read(1)[0]
+        neighbor = self.neighbors_data_set.read(1)[0]
+        true_neighbors = list(map(str, neighbor[:self.k]))
+        self.query_params.update({
+            "neighbors": true_neighbors,
+            "request-params": {
+                "_source": "false",
```
Comment (jmazanec15) [inline osbenchmark/workload/params.py:946]: Should request-params be configurable?

File: osbenchmark/workload/params.py [hunk 3]
```diff
@@ -797,6 +799,183 @@ def params(self):
...
+        return partition_x
+
+    @abstractmethod
+    def params(self):
+        """
+        Returns: A single parameter from this source
+        """
+
+
+class VectorSearchPartitionPartitionParamSource(VectorDataSetPartitionParamSource):
```
Comment (jmazanec15) [inline osbenchmark/workload/params.py:892]: This name seems strange. Why are there 2 partitions? Shouldnt it be `VectorDataSetSearchPartitionParamSource`?

Comment (VijayanB) [reply osbenchmark/workload/params.py:892]: Ack

File: osbenchmark/workload/params.py (hunk 1 above)
Comment (VijayanB) [reply osbenchmark/workload/params.py]: updated to true

File: osbenchmark/workload/params.py (hunk 2 above)
Comment (VijayanB) [reply osbenchmark/workload/params.py]: Moved it to defaults if it is not already available.

File: osbenchmark/workload/params.py [hunk 4]
```diff
@@ -797,6 +799,195 @@ def params(self):
...
+        return partition_x
+
+    @abstractmethod
+    def params(self):
+        """
+        Returns: A single parameter from this source
+        """
+
+
+class VectorSearchPartitionParamSource(VectorDataSetPartitionParamSource):
```
Comment (navneet1v) [inline osbenchmark/workload/params.py:895]: what is the final query that is sent to the Opensearch with this?

**Review by navneet1v — APPROVED** [review]

**Review by jmazanec15 — APPROVED** [review]

File: osbenchmark/workload/params.py [hunk 5]
```diff
@@ -797,6 +799,200 @@ def params(self):
...
+class VectorSearchParamSource(SearchParamSource):
+    def __init__(self, workload, params, **kwargs):
+        super().__init__(workload, params, **kwargs)
+        self.delegate_param_source = VectorSearchPartitionParamSource(params, self.query_params)
+
+    def partition(self, partition_index, total_partitions):
+        return self.delegate_param_source.partition(partition_index, total_partitions)
+
+    def params(self):
+        raise exceptions.BenchmarkError("Do not use a VectorSearchParamSource without partitioning")
```
Comment (IanHoang) [inline osbenchmark/workload/params.py:812]: Might be more appropriate to use `exceptions.WorkloadConfigError` as apposed to catch-all exception `exceptions.BenchmarkError`

**Review by IanHoang — APPROVED** [review]

File: osbenchmark/workload/params.py (hunk 5 above)
Comment (VijayanB) [reply osbenchmark/workload/params.py]: Updated.

**Pull request merged**